db/rocksdb/db.go

```diff
@@ -0,0 +1,353 @@
+func NewDB(dir string) (*dbManager, error) {
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
+	}
+
+	// default rocksdb option, good enough for most cases, including heavy workloads.
+	// 1GB table cache, 512MB write buffer(may use 50% more on heavy workloads).
+	// compression: snappy as default, need to -lsnappy to enable.
+	bbto := gorocksdb.NewDefaultBlockBasedTableOptions()
+	bbto.SetBlockCache(gorocksdb.NewLRUCache(1 << 30))
+	bbto.SetFilterPolicy(gorocksdb.NewBloomFilter(10))
+	dbo := gorocksdb.NewDefaultOptions()
+	dbo.SetBlockBasedTableFactory(bbto)
+	dbo.SetCreateIfMissing(true)
+	dbo.IncreaseParallelism(runtime.NumCPU())
+	// 1.5GB maximum memory use for writebuffer.
+	dbo.OptimizeLevelStyleCompaction(512 * 1024 * 1024)
+
+	opts := dbOptions{
+		dbo: dbo,
+		txo: gorocksdb.NewDefaultOptimisticTransactionOptions(),
+		ro:  gorocksdb.NewDefaultReadOptions(),
+		wo:  gorocksdb.NewDefaultWriteOptions(),
+	}
+	mgr := &dbManager{
+		dir:  dir,
+		opts: opts,
+		mtx:  &sync.RWMutex{},
+	}
+
+	err := os.MkdirAll(mgr.checkpointsDir(), 0755)
+	if err != nil {
+		return nil, err
+	}
+	if mgr.vmgr, err = readVersions(mgr.checkpointsDir()); err != nil {
+		return nil, err
+	}
+	dbPath := filepath.Join(dir, "current.db")
+	mgr.current, err = gorocksdb.OpenOptimisticTransactionDb(dbo, dbPath)
+	if err != nil {
+		return nil, err
+	}
+	return mgr, nil
+}
+
+func (mgr *dbManager) checkpointsDir() string {
+	return filepath.Join(mgr.dir, "checkpoints")
+}
+
+// Reads directory for checkpoints files
+func readVersions(dir string) (*dbm.VersionManager, error) {
+	files, err := os.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	var versions []uint64
+	for _, f := range files {
+		var version uint64
+		if _, err := fmt.Sscanf(f.Name(), checkpointFileFormat, &version); err != nil {
+			return nil, err
+		}
+		versions = append(versions, version)
+	}
+	return dbm.NewVersionManager(versions), nil
+}
+
+func (mgr *dbManager) checkpointPath(ver uint64) (string, error) {
+	dbPath := filepath.Join(mgr.checkpointsDir(), fmt.Sprintf(checkpointFileFormat, ver))
+	if stat, err := os.Stat(dbPath); err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			err = dbm.ErrVersionDoesNotExist
+		}
+		return "", err
+	} else if !stat.IsDir() {
+		return "", dbm.ErrVersionDoesNotExist
+	}
+	return dbPath, nil
+}
+
+func (mgr *dbManager) openCheckpoint(ver uint64) (*dbConnection, error) {
+	dbPath, err := mgr.checkpointPath(ver)
+	if err != nil {
+		return nil, err
+	}
+	db, err := gorocksdb.OpenOptimisticTransactionDb(mgr.opts.dbo, dbPath)
+	if err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+func (mgr *dbManager) Reader() dbm.DBReader {
+	mgr.mtx.RLock()
+	defer mgr.mtx.RUnlock()
+	return &dbTxn{
+		txn: mgr.current.TransactionBegin(mgr.opts.wo, mgr.opts.txo, nil),
+		mgr: mgr,
+	}
+}
+
+func (mgr *dbManager) ReaderAt(ver uint64) (dbm.DBReader, error) {
+	mgr.mtx.RLock()
+	defer mgr.mtx.RUnlock()
+	// TODO: cache opened checkpoints
```

Comment:
Can we either elaborate on this and reference an issue or just remove it if applicable?

Reply:
Yes, sorry. My thought was this would be a useful optimization only if past versions are accessed repeatedly, so I wasn't sure whether to add it. But after reviewing, this is actually needed to allow multiple transactions to coexist, as a lock is created on the checkpoint. So I will implement this now.